File: javascript/src/js/heap/minHeap.js
class MinHeap {

    constructor(array) {
        this.array = this.buildHeap(array);
        this.length = array.length;
    }

    buildHeap(array) {
        var parent = Math.floor((array.length-1)/2)
        while(parent >= 0) {
            this.shiftDown(parent);
            parent--;
        }
    }

    insert(value) {
        array.push(value);
        this.shiftUp(array.length-1)
    }
    
    shiftDown(i) {
        // var parent = Math.floor((i-1)/2) -1 ;
        
        while(i < array.length) {
            var left =   (2*i) + 1;
            var right =  (2*i) + 2;
            var minIndex = -1
            if(left >= array.length || right >= array.length) {
                return;
            }
            if(array[left] < array[right]) {
                minIndex = left;
            } else {
                minIndex = right;
            }

            if(array[i] > array[minIndex]) {
                this.swap(array,i,minIndex)
            } else {
                break;
            }

            i = minIndex;
        }
    }

    shiftUp(i) {
         var parent = Math.floor((i-1)/2) ;
         while(parent >= 0 && array[i] < array[parent]) {
             this.swap(array,i,parent);
             i = parent;
             parent = Math.floor((i-1)/2);
         }
          
    }

    print() {
        console.log(array);
    }

    swap(arr,i,j) {
        var temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
    remove() {
        this.swap(array,0,array.length-1);
        array.pop();
        this.shiftDown(0)
    }
    peek() {
        console.log(array);
        if(array.length > 0) {
            console.log("Peak value =>", array[0]);
            return array[0];
        }  
            
        return -1
      }
   
}
  //   49
 // 12   53
//13  1 4 45

//       1
//   2       4
// 9   5      13 12 
//45 53

// {"arguments": [76], "method": "insert"},
// {"arguments": [], "method": "peek"},
// {"arguments": [], "method": "remove"},
// {"arguments": [], "method": "peek"},
// {"arguments": [], "method": "remove"},
// {"arguments": [], "method": "peek"},
// {"arguments": [87], "method": "insert"}

// let array = [9,5,12,53,1,13,4,45,2]
// let array = [48, 12, 24, 7, 8, -5, 24, 391, 24, 56, 2, 6, 8, 41]
let array = [
    544,
    -578,
    556,
    713,
    -655,
    -359,
    -810,
    -731,
    194,
    -531,
    -685,
    689,
    -279,
    -738,
    886,
    -54,
    -320,
    -500,
    738,
    445,
    -401,
    993,
    -753,
    329,
    -396,
    -924,
    -975,
    376,
    748,
    -356,
    972,
    459,
    399,
    669,
    -488,
    568,
    -702,
    551,
    763,
    -90,
    -249,
    -45,
    452,
    -917,
    394,
    195,
    -877,
    153,
    153,
    788,
    844,
    867,
    266,
    -739,
    904,
    -154,
    -947,
    464,
    343,
    -312,
    150,
    -656,
    528,
    61,
    94,
    -581
  ]
let heap = new MinHeap(array);
heap.buildHeap(array)
console.log(array);
heap.peek();
// heap.shiftDown();
// heap.insert(76);
// heap.print()
// heap.peek();
// heap.remove()
// heap.print()
// heap.peek()
// heap.remove()
// heap.peek()
// heap.insert(87);
// heap.print()

File: java/src/main/java/com/algorithms/algoExpert/MinHeap.java
package com.algorithms.algoExpert;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class MinHeap {

    public static void main(String[] args) {

        int [] arr = {4,6,2,8,9,1,7,-3};
        ArrayList<Integer> list = Arrays.stream(arr).mapToObj(i -> Integer.valueOf(i)).collect(Collectors.toCollection(ArrayList::new));
        System.out.println(list);
        buildHeap(list);
        System.out.println(list);
    }

    public static void buildHeap(ArrayList<Integer> heap) {
        int parent = (heap.size()-1)/2;

        int i = parent;
        while(i >= 0 ) {
            System.out.println("------end----------");
            shiftDown(heap,i,heap.size()-1);
            i--;
            System.out.println("------start----------");
        }

    }

    public static void shiftDown(ArrayList<Integer> heap, int currentIdx, int endIdx) {


        while(true) {
            int leftChild = (currentIdx * 2) + 1;
            int rightChild = (currentIdx * 2) + 2;

            //only one childNode
            if(leftChild <= endIdx && rightChild > endIdx) {
                if(heap.get(leftChild) < heap.get(currentIdx)) {
                    Collections.swap(heap,leftChild,currentIdx);
                } else {
                    break;
                }
                currentIdx = leftChild;
            } else if(leftChild <= endIdx && rightChild <= endIdx) {
                //both children are there
                int pos = -1;
                if(heap.get(leftChild) < heap.get(rightChild)) {
                    pos = leftChild;
                } else {
                    pos = rightChild;
                }

                if(heap.get(pos) < heap.get(currentIdx)) {
                    Collections.swap(heap,pos,currentIdx);
                }
                currentIdx = pos;
            } else {
                // elements greater than the index location
                break;
            }

        }
        System.out.println(heap);
    }


//    public static ArrayList<Integer> buildHeap(ArrayList<Integer> array){
//
//        int i = getParent(array.size()-1);
//        while(i >= 0) {
//            shiftDown(i,array.size()-1,array);
//            // p(array);
//            i--;
//        }
//        System.out.println("Build MaxHeap - > " +array);
//        return null;
//    }

//   public static void shiftDown(int currentIdx,int endIdx, ArrayList<Integer> heap) {

//     int left = getLeftChild(currentIdx);
//     int right = getRightChild(currentIdx);

//          System.out.println(currentIdx +  " = " + endIdx + " = " + left + " = "+ right);


//   if(left <= endIdx && right > endIdx && heap.get(currentIdx) >     heap.get(left) ) {
//       Collections.swap(heap,left,currentIdx);
//     return;
//    }

//      while(left <= endIdx && right <= endIdx) {

//        int pos = 0;
//        if(heap.get(left) < heap.get(right)) {
//          pos = left;
//        } else {
//          pos = right;
//        }

//        if(heap.get(pos) < heap.get(currentIdx)) {
//          Collections.swap(heap,pos,currentIdx);
//        }
//        currentIdx = pos;
//        left = getLeftChild(currentIdx);
//        right = getRightChild(currentIdx);
//     }

//   if(left <= endIdx && right > endIdx && heap.get(currentIdx) >     heap.get(left) ) {
//       Collections.swap(heap,left,currentIdx);
//    }



//   }

    public static void shiftDown(int currentIdx,int endIdx, ArrayList<Integer> heap) {

        int left = getLeftChild(currentIdx);
        int right = getRightChild(currentIdx);

        while(left <= endIdx && right <= endIdx) {
            int pos = 0;
            if(heap.get(left) < heap.get(right)) {
                pos = left;
            } else {
                pos = right;
            }

            if(heap.get(pos) < heap.get(currentIdx)) {
                Collections.swap(heap,pos,currentIdx);
            }
            currentIdx = pos;
            left = getLeftChild(currentIdx);
            right = getRightChild(currentIdx);
        }

        System.out.println(heap);

    }


    public static void shiftUp(int currentIdx, ArrayList<Integer> heap) {

        int parent =  getParent(currentIdx);
        while(parent >= 0 && heap.get(parent) > heap.get(currentIdx)) {
            Collections.swap(heap,parent,currentIdx);
            currentIdx = parent;
            parent = getParent(currentIdx);
        }

    }

//    public static int peek() {
//        if(list.isEmpty()){
//            return -1;
//        }
//        return list.get(0);
//    }
//
//    public static int remove() {
//        int val = list.get(0);
//        Collections.swap(list,0,list.size()-1);
//        list.remove(list.size()-1);
//        shiftDown(0,list.size()-1,list);
//        System.out.println("Remove -> " + list);
//        return val;
//    }
//
//    public static void insert(int value) {
//        list.add(value);
//        shiftUp(list.size()-1,list);
//        System.out.println("Insert -> " + list);
//
//    }


    public static void p(ArrayList<Integer> list){
        System.out.println(list);
    }

    private static int getParent(int currentIdx) {
        Double floor = Math.floor((currentIdx-1)/2);
        return floor.intValue();
    }

    public static int getLeftChild(int currentIdx) {
        return (2*currentIdx)+1;
    }


    public static int getRightChild(int currentIdx) {
        return (2*currentIdx)+2;
    }


}
